target_chains/ton/contracts/contracts/Pyth.fc

```diff
@@ -74,3 +102,67 @@ slice verify_header(slice data) {
     throw_if(ERROR_OUTDATED_PRICE, current_time - publish_time > time_period);
     return (price, conf, expo, publish_time);
 }
+
+() update_price_feeds(int msg_value, slice data) impure {
+    load_data();
+    slice cs = read_and_verify_header(data);
+
+    int wormhole_proof_size_bytes = cs~load_uint(16);
+    (cell wormhole_proof, slice new_cs) = read_and_store_large_data(cs, wormhole_proof_size_bytes * 8);
+    cs = new_cs;
+
+    int num_updates = cs~load_uint(8);
+    int fee = single_update_fee * num_updates;
+
+    ;; Check if the sender has sent enough TON to cover the fee
+    throw_unless(ERROR_INSUFFICIENT_FEE, msg_value >= fee);
+
+    (_, _, _, _, int emitter_chain_id, int emitter_address, _, _, slice payload, _) = parse_and_verify_wormhole_vm(wormhole_proof.begin_parse());
+
+    ;; Check if the data source is valid
+    cell data_source = begin_cell()
+        .store_uint(emitter_chain_id, 16)
+        .store_uint(emitter_address, 256)
+        .end_cell();
+
+    ;; Dictionary doesn't support cell as key, so we use cell_hash to create a 256-bit integer key
+    int data_source_key = cell_hash(data_source);
+
+    (slice value, int found?) = is_valid_data_source.udict_get?(256, data_source_key);
+    throw_unless(ERROR_UPDATE_DATA_SOURCE_NOT_FOUND, found?);
+    int valid = value~load_int(1);
+    throw_unless(ERROR_INVALID_UPDATE_DATA_SOURCE, (valid == -1));
+
+
+    int root_digest = parse_wormhole_payload(payload);
+
+    repeat(num_updates) {
+        (int price_id, int price, int conf, int expo, int publish_time, int prev_publish_time, int ema_price, int ema_conf, slice new_cs) = read_and_verify_message(cs, root_digest);
+        cs = new_cs;
+
+        (slice latest_price_info, int found?) = latest_price_feeds.udict_get?(256, price_id);
+        int latest_publish_time = 0;
+        if (found?) {
+            slice price_feed_slice = latest_price_info~load_ref().begin_parse();
+            slice price_slice = price_feed_slice~load_ref().begin_parse();
+
+            price_slice~load_int(64);  ;; Skip price
+            price_slice~load_uint(64); ;; Skip conf
+            price_slice~load_int(32);  ;; Skip expo
+            latest_publish_time = price_slice~load_uint(64);
```

Comment:
nit: i think having latest_publish_time as the first element is more ergonomic in the places it is being used.

Reply:
hmm i was trying to keep it as the same order as the message in `read_and_verify_message`